File: examples/blockchainnews-agent/src/service.py
"""
Blockchain News Agent — APEX (ERC-8183) Provider.

Built with bnbagent-sdk.

A news search agent that:
  1. Receives search queries from clients via APEX
  2. Searches DuckDuckGo for blockchain news
  3. Returns formatted news results

Usage:
    cd agents
    uv run python -m blockchain_news.service

Environment (blockchain_news/.env):
    RPC_URL, ERC8183_ADDRESS, PRIVATE_KEY      — Required (ERC-8183)
    APEX_EVALUATOR_ADDRESS                     — Required (evaluator)
    STORAGE_PROVIDER=ipfs, PINATA_JWT          — Required (IPFS upload)
    AGENT_PRICE=1000000000000000000            — Negotiation price (1 U)
    PAYMENT_TOKEN_ADDRESS                      — BEP20 payment token
    PORT=8003                                  — Server port
    POLL_INTERVAL=15                           — Job polling interval
"""

import asyncio
import json
import logging
import os
from pathlib import Path

from dotenv import load_dotenv
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
from ddgs import DDGS

# Load .env from project root (one level up from src/)
load_dotenv(Path(__file__).resolve().parent.parent / ".env")

# SDK imports — same pattern as bnbagent-sdk/examples/full_agent.py
from bnbagent.quickstart import APEXConfig, APEXState, create_apex_state, create_apex_routes

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(name)s] %(levelname)s: %(message)s",
)
logger = logging.getLogger("blockchain_news")

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

config = APEXConfig.from_env()
state: APEXState = create_apex_state(config)

PORT = int(os.getenv("PORT", "8003"))
POLL_INTERVAL = int(os.getenv("POLL_INTERVAL", "15"))

logger.info(f"APEX enabled: agent={state.job_ops.agent_address}")

# ---------------------------------------------------------------------------
# Pydantic models for direct /search endpoint
# ---------------------------------------------------------------------------


class SearchRequest(BaseModel):
    query: str
    max_results: int = 10


class NewsItem(BaseModel):
    title: str
    body: str
    url: str
    date: str
    source: str


class SearchResponse(BaseModel):
    success: bool
    query: str
    results_count: int
    results: list[NewsItem]


# ---------------------------------------------------------------------------
# App — SDK quickstart routes mounted on FastAPI
# ---------------------------------------------------------------------------

app = FastAPI(
    title="Blockchain News Agent",
    description="News search service with ERC-8183 APEX protocol support",
)
app.include_router(create_apex_routes(state=state), prefix="/apex")


# ---------------------------------------------------------------------------
# Core news search function
# ---------------------------------------------------------------------------


def search_news(query: str, max_results: int = 10) -> list[dict]:
    """Search news using DuckDuckGo."""
    ddgs = DDGS()

    results = list(ddgs.news(query, max_results=max_results))
    if not results:
        results = list(ddgs.text(query, max_results=max_results))

    return results


def format_news_results(query: str, raw_results: list[dict]) -> str:
    """Format news results into a readable report."""
    if not raw_results:
        return f"No news found for query: {query}"

    report = f"# Blockchain News Search Results\n\n"
    report += f"**Query:** {query}\n"
    report += f"**Results:** {len(raw_results)} items\n\n"
    report += "---\n\n"

    for i, r in enumerate(raw_results, 1):
        title = r.get("title", "No title")
        body = r.get("body", r.get("snippet", ""))
        url = r.get("url", r.get("href", ""))
        date = r.get("date", "")
        source = r.get("source", "")

        report += f"## {i}. {title}\n\n"
        if source or date:
            report += f"*{source}*"
            if date:
                report += f" | {date}"
            report += "\n\n"
        report += f"{body}\n\n"
        if url:
            report += f"[Read more]({url})\n\n"
        report += "---\n\n"

    return report


# ---------------------------------------------------------------------------
# Direct HTTP endpoint (for testing without APEX)
# ---------------------------------------------------------------------------


@app.post("/search", response_model=SearchResponse)
async def search_endpoint(request: SearchRequest):
    """
    Direct HTTP search endpoint (for testing).
    For production, use APEX protocol via /apex/* endpoints.
    """
    try:
        raw_results = search_news(request.query, request.max_results)

        results = []
        for r in raw_results:
            results.append(
                NewsItem(
                    title=r.get("title", ""),
                    body=r.get("body", r.get("snippet", "")),
                    url=r.get("url", r.get("href", "")),
                    date=r.get("date", ""),
                    source=r.get("source", ""),
                )
            )

        return SearchResponse(
            success=True,
            query=request.query,
            results_count=len(results),
            results=results,
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@app.get("/health")
async def health():
    return {
        "status": "ok",
        "agent": "blockchain-news",
        "apex_enabled": True,
        "agent_address": state.job_ops.agent_address,
    }


# ---------------------------------------------------------------------------
# APEX Task Processing
# ---------------------------------------------------------------------------


async def process_task(query: str) -> str:
    """Process a news search task from APEX."""
    logger.info(f"Searching news for: {query[:80]}...")

    raw_results = search_news(query, max_results=10)
    logger.info(f"Found {len(raw_results)} news items")

    report = format_news_results(query, raw_results)
    return report


# ---------------------------------------------------------------------------
# Background polling loop — verify → process → submit
# ---------------------------------------------------------------------------


async def poll_funded_jobs():
    """Poll for FUNDED jobs assigned to this agent and process them."""
    logger.info(f"Polling for funded jobs every {POLL_INTERVAL}s...")
    my_address = state.job_ops.agent_address
    logger.info(f"My address: {my_address}")

    while True:
        try:
            logger.info("Polling for funded jobs...")
            result = await state.job_ops.get_pending_jobs()

            if not result.get("success"):
                logger.warning(f"get_pending_jobs error: {result.get('error')}")
                await asyncio.sleep(POLL_INTERVAL)
                continue

            jobs = result.get("jobs", [])
            logger.info(f"Found {len(jobs)} pending job(s)")

            for job in jobs:
                job_id = job["jobId"]
                description = job.get("description", "")
                logger.info(f"Found FUNDED job #{job_id}: {description[:60]}...")

                verification = await state.job_ops.verify_job(job_id)
                if not verification["valid"]:
                    logger.warning(
                        f"Job #{job_id} verification failed: {verification.get('error')}"
                    )
                    continue

                try:
                    news_report = await process_task(description)
                except Exception as e:
                    logger.error(f"Task processing failed for job #{job_id}: {e}")
                    continue

                logger.info(f"Submitting result for job #{job_id}...")
                submission = await state.job_ops.submit_result(
                    job_id=job_id,
                    response_content=news_report,
                    metadata={
                        "agent": "blockchain-news",
                        "query": description,
                    },
                )

                if submission.get("success"):
                    logger.info(f"Job #{job_id} submitted! TX: {submission['txHash']}")
                    if submission.get("dataUrl"):
                        logger.info(f"  IPFS: {submission['dataUrl']}")
                else:
                    logger.error(
                        f"Job #{job_id} submission failed: {submission.get('error')}"
                    )

        except Exception as e:
            logger.error(f"Polling error: {e}")

        await asyncio.sleep(POLL_INTERVAL)


@app.on_event("startup")
async def startup():
    asyncio.create_task(poll_funded_jobs())


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import uvicorn

    print(f"""
{'='*55}
  Blockchain News Agent (APEX Provider)
{'='*55}
  Port:           {PORT}
  Agent Address:  {state.job_ops.agent_address}
  ERC-8183:       {config.erc8183_address}
  Evaluator:      {config.apex_evaluator_address}
  Storage:        {config.storage_provider}
  Price:          {int(config.agent_price) / 10**18} U tokens

  APEX endpoints:
    POST /apex/negotiate   — Negotiation
    POST /apex/submit      — Submit result
    GET  /apex/job/{{id}}    — Job details
    GET  /apex/status      — Agent status

  Direct endpoints (testing):
    POST /search          — Direct news search
    GET  /health          — Health check

  Polling for funded jobs every {POLL_INTERVAL}s...
{'='*55}
""")

    uvicorn.run(app, host="0.0.0.0", port=PORT)

File: typescript/examples/agent-server/src/service.ts
/**
 * Blockchain News Agent — ERC-8183 provider (HTTP serving reference).
 *
 * TypeScript port of `python/examples/agent-server/src/service.py`. A news
 * search agent that:
 *   1. receives funded jobs via ERC-8183 (the funded-job poll loop),
 *   2. searches DuckDuckGo for blockchain news,
 *   3. returns a formatted report the SDK submits on-chain.
 *
 * Unlike the Python reference there is no TWAK wallet switch — the TypeScript
 * SDK ships an EVM wallet provider only, so this uses `ERC8183Config.fromEnv()`
 * (EVM keystore) end to end.
 *
 * Run:
 *   pnpm -C typescript example:agent-server
 *   # or: pnpm -C typescript exec tsx examples/agent-server/src/service.ts
 *
 * Env (examples/agent-server/.env — see .env.example):
 *   WALLET_PASSWORD, PRIVATE_KEY (first run), NETWORK, ERC8183_AGENT_URL,
 *   ERC8183_SERVICE_PRICE, PORT (default 8003), ERC8183_FUNDED_POLL_INTERVAL, …
 */

import { dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { ERC8183Config } from "../../../src/erc8183/index.js";
import { loadEnv } from "../../../src/index.js";
import { LocalStorageProvider } from "../../../src/storage/index.js";
import { createErc8183Server } from "./erc8183Server.js";
import { route, sendJson } from "./http.js";

const ROOT = dirname(dirname(fileURLToPath(import.meta.url))); // examples/agent-server
loadEnv(ROOT);

// ── Storage backend — pick ONE ──
// (a) Local filesystem (default). ERC8183_AGENT_URL is required: the SDK
//     rewrites file:// URLs to this public base URL.
const storage = LocalStorageProvider.fromEnv();
// (b) IPFS via Pinata — set STORAGE_API_KEY (Pinata JWT), then swap to:
// import { IPFSStorageProvider } from "../../../src/storage/index.js";
// const storage = IPFSStorageProvider.fromEnv();

const PORT = Number(process.env.PORT ?? "8003");

// ---------------------------------------------------------------------------
// News search (DuckDuckGo lite over global fetch — no runtime dependency)
// ---------------------------------------------------------------------------

interface NewsItem {
  title: string;
  body: string;
  url: string;
}

function stripTags(html: string): string {
  return html
    .replace(/<[^>]+>/g, "")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&#x27;/g, "'")
    .replace(/&quot;/g, '"')
    .replace(/\s+/g, " ")
    .trim();
}

/**
 * Best-effort blockchain-news search via DuckDuckGo's dependency-free "lite"
 * HTML endpoint. Returns `[]` on any failure (network, parse, empty) so the
 * agent still produces a deliverable rather than stranding a funded job.
 */
async function searchNews(query: string, maxResults = 10): Promise<NewsItem[]> {
  try {
    const resp = await fetch("https://lite.duckduckgo.com/lite/", {
      method: "POST",
      headers: {
        "content-type": "application/x-www-form-urlencoded",
        "user-agent": "Mozilla/5.0 (compatible; bnbagent-example/1.0)",
      },
      body: new URLSearchParams({ q: query }).toString(),
      signal: AbortSignal.timeout(15_000),
    });
    if (!resp.ok) {
      return [];
    }
    const html = await resp.text();
    const linkRe =
      /<a[^>]*class="result-link"[^>]*href="([^"]+)"[^>]*>([\s\S]*?)<\/a>/g;
    const snippetRe = /<td[^>]*class="result-snippet"[^>]*>([\s\S]*?)<\/td>/g;
    const snippets = [...html.matchAll(snippetRe)].map((m) =>
      stripTags(m[1] ?? ""),
    );
    const items: NewsItem[] = [];
    let i = 0;
    for (const m of html.matchAll(linkRe)) {
      if (items.length >= maxResults) {
        break;
      }
      items.push({
        url: m[1] ?? "",
        title: stripTags(m[2] ?? ""),
        body: snippets[i] ?? "",
      });
      i += 1;
    }
    return items;
  } catch (error) {
    console.warn(
      `[blockchain-news] search failed: ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
    return [];
  }
}

function formatNewsResults(query: string, items: NewsItem[]): string {
  if (items.length === 0) {
    return `No news found for query: ${query}`;
  }
  const lines = [
    "# Blockchain News Search Results",
    "",
    `**Query:** ${query}`,
    `**Results:** ${items.length} items`,
    "",
    "---",
    "",
  ];
  items.forEach((item, idx) => {
    lines.push(`## ${idx + 1}. ${item.title}`, "");
    if (item.body) {
      lines.push(item.body, "");
    }
    if (item.url) {
      lines.push(`[Read more](${item.url})`, "");
    }
    lines.push("---", "");
  });
  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// The ERC-8183 task handler — the ONLY function you need to write.
// The SDK calls this for each funded job; return (report, metadata).
// ---------------------------------------------------------------------------

async function processTask(
  job: Record<string, unknown>,
): Promise<[string, Record<string, unknown>]> {
  const { parseJobDescription } = await import("../../../src/erc8183/index.js");
  const rawDescription = String(job.description ?? "blockchain news");
  const parsed = parseJobDescription(rawDescription);
  const query = parsed?.task ?? rawDescription;
  console.info(`[blockchain-news] searching news for: ${query.slice(0, 80)}…`);

  const results = await searchNews(query, 10);
  console.info(`[blockchain-news] found ${results.length} news items`);

  return [
    formatNewsResults(query, results),
    { agent: "blockchain-news", query },
  ];
}

// ---------------------------------------------------------------------------
// Boot
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const config = ERC8183Config.fromEnv(storage);

  // Direct /search endpoint for testing without ERC-8183.
  const searchRoute = route("POST", "/search", async ({ res, body }) => {
    const req = (body ?? {}) as { query?: string; max_results?: number };
    if (!req.query) {
      return sendJson(res, 400, { error: "query is required" });
    }
    const results = await searchNews(req.query, req.max_results ?? 10);
    sendJson(res, 200, {
      success: true,
      query: req.query,
      results_count: results.length,
      results,
    });
  });

  const server = await createErc8183Server({
    config,
    onJob: processTask,
    extraRoutes: [searchRoute],
  });

  console.log(`
${"=".repeat(55)}
  Blockchain News Agent (ERC-8183 Provider)
${"=".repeat(55)}
  Port:       ${PORT}
  Agent:      ${server.agentAddress}
  Commerce:   ${config.effectiveCommerceAddress}
  Router:     ${config.effectiveRouterAddress}
  Policy:     ${config.effectivePolicyAddress}
  Storage:    ${storage.constructor.name}
  Price:      ${Number(config.servicePrice) / 1e18} tokens

  ERC-8183 endpoints:
    POST /erc8183/negotiate      — Negotiation
    GET  /erc8183/job/{id}       — Job details
    GET  /erc8183/status         — Agent status
  Direct (testing):
    POST /search                 — Direct news search
    GET  /erc8183/health         — Health check
${"=".repeat(55)}
`);

  server.listen(PORT);
  console.info(`[blockchain-news] listening on http://localhost:${PORT}`);
}

main().catch((error) => {
  console.error(error instanceof Error ? error.stack : error);
  process.exit(1);
});
